# Sales Data Assignment — Python/pandas Version
- https://docs.google.com/spreadsheets/d/1tTEalsKYhP_Wl8-ddSx6opUAccwqbt2OzWWzEnr8pIc/edit?gid=1652222712#gid=1652222712
- The workbook contains 25 sales rows, including a duplicated O1024 record, and 20 tasks covering aggregation, filtering, conditional formatting, lookups, pivoting, sorting, percentages, and classification.
- We will convert each task into its pandas/Python equivalent and validating the results
- This notebook converts the original Google Sheets/Excel assignment into Python exercises and provides complete solutions.

**Dataset:** 25 sales records and 8 columns. The data intentionally contains a duplicate `OrderID` (`O1024`) so duplicate detection can be practised.

## Learning objectives
- Create and inspect a pandas DataFrame
- Filter and sort data
- Calculate totals, averages, counts, conditional summaries and percentages
- Detect duplicates
- Perform lookup operations
- Create pivot-style summaries
- Apply conditional styling
- Protect the original data by working on copies


## Setup and DataFrame creation

In [8]:
import pandas as pd
import numpy as np
from pathlib import Path

pd.set_option('display.max_columns', None)
pd.set_option('display.float_format', lambda x: f'{x:,.2f}')

In [10]:
%whoS #display the variables currently available

No variables match your requested type.


### Create the DataFrame directly from the assignment data

In [5]:
sales_data = [{'OrderID': 'O1001', 'Salesperson': 'Asha', 'Region': 'North', 'Product': 'Laptop', 'Category': 'Electronics', 'Units': 5, 'UnitPrice': 500, 'Revenue': 2500}, {'OrderID': 'O1002', 'Salesperson': 'Ravi', 'Region': 'South', 'Product': 'Mouse', 'Category': 'Accessories', 'Units': 20, 'UnitPrice': 20, 'Revenue': 400}, {'OrderID': 'O1003', 'Salesperson': 'Neha', 'Region': 'West', 'Product': 'Keyboard', 'Category': 'Accessories', 'Units': 15, 'UnitPrice': 30, 'Revenue': 450}, {'OrderID': 'O1004', 'Salesperson': 'Amit', 'Region': 'East', 'Product': 'Monitor', 'Category': 'Electronics', 'Units': 7, 'UnitPrice': 180, 'Revenue': 1260}, {'OrderID': 'O1005', 'Salesperson': 'Asha', 'Region': 'North', 'Product': 'Printer', 'Category': 'Office', 'Units': 3, 'UnitPrice': 220, 'Revenue': 660}, {'OrderID': 'O1006', 'Salesperson': 'Ravi', 'Region': 'West', 'Product': 'Laptop', 'Category': 'Electronics', 'Units': 2, 'UnitPrice': 500, 'Revenue': 1000}, {'OrderID': 'O1007', 'Salesperson': 'Neha', 'Region': 'South', 'Product': 'Chair', 'Category': 'Furniture', 'Units': 6, 'UnitPrice': 90, 'Revenue': 540}, {'OrderID': 'O1008', 'Salesperson': 'Amit', 'Region': 'East', 'Product': 'Desk', 'Category': 'Furniture', 'Units': 2, 'UnitPrice': 200, 'Revenue': 400}, {'OrderID': 'O1009', 'Salesperson': 'Asha', 'Region': 'West', 'Product': 'Mouse', 'Category': 'Accessories', 'Units': 25, 'UnitPrice': 20, 'Revenue': 500}, {'OrderID': 'O1010', 'Salesperson': 'Ravi', 'Region': 'North', 'Product': 'Monitor', 'Category': 'Electronics', 'Units': 4, 'UnitPrice': 180, 'Revenue': 720}, {'OrderID': 'O1011', 'Salesperson': 'Neha', 'Region': 'East', 'Product': 'Laptop', 'Category': 'Electronics', 'Units': 3, 'UnitPrice': 500, 'Revenue': 1500}, {'OrderID': 'O1012', 'Salesperson': 'Amit', 'Region': 'South', 'Product': 'Printer', 'Category': 'Office', 'Units': 4, 'UnitPrice': 220, 'Revenue': 880}, {'OrderID': 'O1013', 'Salesperson': 'Asha', 'Region': 'North', 'Product': 'Keyboard', 'Category': 'Accessories', 'Units': 10, 'UnitPrice': 30, 'Revenue': 300}, {'OrderID': 'O1014', 'Salesperson': 'Ravi', 'Region': 'East', 'Product': 'Desk', 'Category': 'Furniture', 'Units': 1, 'UnitPrice': 200, 'Revenue': 200}, {'OrderID': 'O1015', 'Salesperson': 'Neha', 'Region': 'West', 'Product': 'Chair', 'Category': 'Furniture', 'Units': 8, 'UnitPrice': 90, 'Revenue': 720}, {'OrderID': 'O1016', 'Salesperson': 'Amit', 'Region': 'North', 'Product': 'Laptop', 'Category': 'Electronics', 'Units': 6, 'UnitPrice': 500, 'Revenue': 3000}, {'OrderID': 'O1017', 'Salesperson': 'Asha', 'Region': 'South', 'Product': 'Monitor', 'Category': 'Electronics', 'Units': 5, 'UnitPrice': 180, 'Revenue': 900}, {'OrderID': 'O1018', 'Salesperson': 'Ravi', 'Region': 'West', 'Product': 'Printer', 'Category': 'Office', 'Units': 2, 'UnitPrice': 220, 'Revenue': 440}, {'OrderID': 'O1019', 'Salesperson': 'Neha', 'Region': 'East', 'Product': 'Mouse', 'Category': 'Accessories', 'Units': 18, 'UnitPrice': 20, 'Revenue': 360}, {'OrderID': 'O1020', 'Salesperson': 'Amit', 'Region': 'South', 'Product': 'Keyboard', 'Category': 'Accessories', 'Units': 14, 'UnitPrice': 30, 'Revenue': 420}, {'OrderID': 'O1021', 'Salesperson': 'Asha', 'Region': 'North', 'Product': 'Desk', 'Category': 'Furniture', 'Units': 2, 'UnitPrice': 200, 'Revenue': 400}, {'OrderID': 'O1022', 'Salesperson': 'Ravi', 'Region': 'West', 'Product': 'Chair', 'Category': 'Furniture', 'Units': 5, 'UnitPrice': 90, 'Revenue': 450}, {'OrderID': 'O1023', 'Salesperson': 'Neha', 'Region': 'South', 'Product': 'Laptop', 'Category': 'Electronics', 'Units': 4, 'UnitPrice': 500, 'Revenue': 2000}, {'OrderID': 'O1024', 'Salesperson': 'Amit', 'Region': 'East', 'Product': 'Monitor', 'Category': 'Electronics', 'Units': 9, 'UnitPrice': 180, 'Revenue': 1620}, {'OrderID': 'O1024', 'Salesperson': 'Amit', 'Region': 'East', 'Product': 'Monitor', 'Category': 'Electronics', 'Units': 9, 'UnitPrice': 180, 'Revenue': 1620}]
df = pd.DataFrame(sales_data)
df.head()

,OrderID,Salesperson,Region,Product,Category,Units,UnitPrice,Revenue
0,O1001,Asha,North,Laptop,Electronics,5,500,2500
1,O1002,Ravi,South,Mouse,Accessories,20,20,400
2,O1003,Neha,West,Keyboard,Accessories,15,30,450
3,O1004,Amit,East,Monitor,Electronics,7,180,1260
4,O1005,Asha,North,Printer,Office,3,220,660


In [12]:
df.isna().sum(axis=0)  #finding count of missing values in each column axis=0 : columns

OrderID        0
Salesperson    0
Region         0
Product        0
Category       0
Units          0
UnitPrice      0
Revenue        0
dtype: int64

In [14]:
df.loc[[2,5],]

,OrderID,Salesperson,Region,Product,Category,Units,UnitPrice,Revenue
2,O1003,Neha,West,Keyboard,Accessories,15,30,450
5,O1006,Ravi,West,Laptop,Electronics,2,500,1000


In [15]:
## insert few missing values in revenue column
df.loc[[2,5],'Revenue'] = pd.NA
df.isna().sum()

OrderID        0
Salesperson    0
Region         0
Product        0
Category       0
Units          0
UnitPrice      0
Revenue        2
dtype: int64

In [16]:
df.loc[[2,5],]

,OrderID,Salesperson,Region,Product,Category,Units,UnitPrice,Revenue
2,O1003,Neha,West,Keyboard,Accessories,15,30,NaN
5,O1006,Ravi,West,Laptop,Electronics,2,500,NaN


### Optional: read the same data directly from the supplied workbook

In [ ]:
file_path = Path('PIIT_SalesData.csv')
df1 = pd.read_csv(file_path)
df1.head()

## 1. Freeze the top row

In pandas there is no scrolling worksheet to freeze. The Python equivalent is to keep column names visible while displaying data and, during Excel export, use `freeze_panes=(1, 0)`. Here we verify the headers.

In [17]:
pd.DataFrame([df.columns])

,0,1,2,3,4,5,6,7
0,OrderID,Salesperson,Region,Product,Category,Units,UnitPrice,Revenue


## 2. Apply a filter and sort Revenue in descending order

Filter-like operations use Boolean indexing; sorting uses `sort_values()`.

In [18]:
df_revenue = df.sort_values('Revenue', ascending=False).reset_index(drop=True)
df_revenue.head()

,OrderID,Salesperson,Region,Product,Category,Units,UnitPrice,Revenue
0,O1016,Amit,North,Laptop,Electronics,6,500,"3,000.00"
1,O1001,Asha,North,Laptop,Electronics,5,500,"2,500.00"
2,O1023,Neha,South,Laptop,Electronics,4,500,"2,000.00"
3,O1024,Amit,East,Monitor,Electronics,9,180,"1,620.00"
4,O1024,Amit,East,Monitor,Electronics,9,180,"1,620.00"


## 3. Calculate Total Revenue

Use the `sum()` method.

In [19]:
total_revenue = df['Revenue'].sum()
total_revenue

np.float64(21790.0)

## 4. Calculate Average Revenue

Use the `mean()` method.

In [20]:
avg_revenue = df['Revenue'].mean()
avg_revenue

np.float64(947.3913043478261)

## 5. Count total records

`len(df)` counts rows, including rows with missing values in some columns.

In [21]:
total_records = len(df)
total_records

25

## 6. Count numeric Revenue cells

`count()` counts non-missing values in the Revenue column.

In [22]:
df.isna().sum()

OrderID        0
Salesperson    0
Region         0
Product        0
Category       0
Units          0
UnitPrice      0
Revenue        2
dtype: int64

In [23]:
nonNA_revenue = df['Revenue'].count()
nonNA_revenue

np.int64(23)

## 7. COUNTIF: Orders from North region

Create a Boolean condition and sum `True` values.

In [24]:
north_orders = df['Region'].eq('North').sum()
north_orders

np.int64(6)

In [25]:
df[df['Region']=='North']
#len(df[df['Region'] == 'North'])

,OrderID,Salesperson,Region,Product,Category,Units,UnitPrice,Revenue
0,O1001,Asha,North,Laptop,Electronics,5,500,"2,500.00"
4,O1005,Asha,North,Printer,Office,3,220,660.00
9,O1010,Ravi,North,Monitor,Electronics,4,180,720.00
12,O1013,Asha,North,Keyboard,Accessories,10,30,300.00
15,O1016,Amit,North,Laptop,Electronics,6,500,"3,000.00"
20,O1021,Asha,North,Desk,Furniture,2,200,400.00


## 8. SUMIF: Total Revenue for Electronics

Filter the category and sum Revenue.

In [26]:
electronics_revenue = df.loc[df['Category'].eq('Electronics'), 'Revenue'].sum()
electronics_revenue

np.float64(15120.0)

## 9. AVERAGEIF: Average Units sold by Asha

Filter the salesperson and calculate the mean Units.

In [27]:
asha_average_units = df.loc[df['Salesperson'].eq('Asha'), 'Units'].mean()
asha_average_units

np.float64(8.333333333333334)

## 10. Highlight Revenue greater than 1,000

Use a pandas `Styler` to apply conditional formatting.

In [28]:
def highlight_high_revenue(value):
    return 'background-color: lightgreen; font-weight: bold' if value > 1000 else ''

df.style.map(highlight_high_revenue, subset=['Revenue']).format({'UnitPrice': '{:,.2f}', 'Revenue': '{:,.2f}'})

,OrderID,Salesperson,Region,Product,Category,Units,UnitPrice,Revenue
0,O1001,Asha,North,Laptop,Electronics,5,500.00,"2,500.00"
1,O1002,Ravi,South,Mouse,Accessories,20,20.00,400.00
2,O1003,Neha,West,Keyboard,Accessories,15,30.00,nan
3,O1004,Amit,East,Monitor,Electronics,7,180.00,"1,260.00"
4,O1005,Asha,North,Printer,Office,3,220.00,660.00
5,O1006,Ravi,West,Laptop,Electronics,2,500.00,nan
6,O1007,Neha,South,Chair,Furniture,6,90.00,540.00
7,O1008,Amit,East,Desk,Furniture,2,200.00,400.00
8,O1009,Asha,West,Mouse,Accessories,25,20.00,500.00
9,O1010,Ravi,North,Monitor,Electronics,4,180.00,720.00


## 11. Apply a colour scale to Revenue

Use `background_gradient()` on the Revenue column.

In [29]:
df.style.background_gradient(subset=['Revenue'], cmap='RdYlGn').format({'UnitPrice': '{:,.2f}', 'Revenue': '{:,.2f}'})

,OrderID,Salesperson,Region,Product,Category,Units,UnitPrice,Revenue
0,O1001,Asha,North,Laptop,Electronics,5,500.00,"2,500.00"
1,O1002,Ravi,South,Mouse,Accessories,20,20.00,400.00
2,O1003,Neha,West,Keyboard,Accessories,15,30.00,nan
3,O1004,Amit,East,Monitor,Electronics,7,180.00,"1,260.00"
4,O1005,Asha,North,Printer,Office,3,220.00,660.00
5,O1006,Ravi,West,Laptop,Electronics,2,500.00,nan
6,O1007,Neha,South,Chair,Furniture,6,90.00,540.00
7,O1008,Amit,East,Desk,Furniture,2,200.00,400.00
8,O1009,Asha,West,Mouse,Accessories,25,20.00,500.00
9,O1010,Ravi,North,Monitor,Electronics,4,180.00,720.00


## 12. Identify duplicate OrderID values

Use `duplicated(keep=False)` to show every occurrence of a repeated ID.

In [30]:
duplicate_orders = df[df['OrderID'].duplicated(keep=False)]
duplicate_orders

,OrderID,Salesperson,Region,Product,Category,Units,UnitPrice,Revenue
23,O1024,Amit,East,Monitor,Electronics,9,180,"1,620.00"
24,O1024,Amit,East,Monitor,Electronics,9,180,"1,620.00"


## 13. Create a lookup table and perform a VLOOKUP-style lookup

In pandas, `map()` or `merge()` replaces VLOOKUP. First create a unique Product–Category lookup table.

In [31]:
products = df['Product'].unique()
products

array(['Laptop', 'Mouse', 'Keyboard', 'Monitor', 'Printer', 'Chair',
       'Desk'], dtype=object)

In [32]:
df[df['Product'].isin(products)][['Product','Category']].drop_duplicates()

,Product,Category
0,Laptop,Electronics
1,Mouse,Accessories
2,Keyboard,Accessories
3,Monitor,Electronics
4,Printer,Office
6,Chair,Furniture
7,Desk,Furniture


In [33]:
#selected products
products1 = ['Chair','Laptop']
df[df['Product'].isin(products1)]

,OrderID,Salesperson,Region,Product,Category,Units,UnitPrice,Revenue
0,O1001,Asha,North,Laptop,Electronics,5,500,"2,500.00"
5,O1006,Ravi,West,Laptop,Electronics,2,500,NaN
6,O1007,Neha,South,Chair,Furniture,6,90,540.00
10,O1011,Neha,East,Laptop,Electronics,3,500,"1,500.00"
14,O1015,Neha,West,Chair,Furniture,8,90,720.00
15,O1016,Amit,North,Laptop,Electronics,6,500,"3,000.00"
21,O1022,Ravi,West,Chair,Furniture,5,90,450.00
22,O1023,Neha,South,Laptop,Electronics,4,500,"2,000.00"


## 14. Repeat using an XLOOKUP-style approach

A left `merge()` is the closest general equivalent of XLOOKUP.

In [34]:
products_df = pd.DataFrame({'Product': products})
products_df

,Product
0,Laptop
1,Mouse
2,Keyboard
3,Monitor
4,Printer
5,Chair
6,Desk


In [35]:
df[['OrderID', 'Product','Category','Salesperson']].merge(products_df, on='Product', how='left', validate='many_to_one')

,OrderID,Product,Category,Salesperson
0,O1001,Laptop,Electronics,Asha
1,O1002,Mouse,Accessories,Ravi
2,O1003,Keyboard,Accessories,Neha
3,O1004,Monitor,Electronics,Amit
4,O1005,Printer,Office,Asha
5,O1006,Laptop,Electronics,Ravi
6,O1007,Chair,Furniture,Neha
7,O1008,Desk,Furniture,Amit
8,O1009,Mouse,Accessories,Asha
9,O1010,Monitor,Electronics,Ravi


## 15. QUERY pivot: Sum of Revenue by Region

Use `groupby()` or `pivot_table()`.

In [36]:
# Groupby - recommended
df.groupby('Region', as_index=False)['Revenue'].sum().sort_values('Revenue', ascending=False)

,Region,Revenue
1,North,"7,580.00"
0,East,"6,960.00"
2,South,"5,140.00"
3,West,"2,110.00"


In [37]:
# Pivot -> aggfunc
df.pivot_table(index='Region', values='Revenue', aggfunc='sum').sort_values('Revenue', ascending=False)

,Revenue
Region,
North,"7,580.00"
East,"6,960.00"
South,"5,140.00"
West,"2,110.00"


## 16. SORT formula: Display data by Units descending

Use `sort_values()`.

In [38]:
df.sort_values(['Units', 'Revenue'], ascending=[False, False]).reset_index(drop=True)

,OrderID,Salesperson,Region,Product,Category,Units,UnitPrice,Revenue
0,O1009,Asha,West,Mouse,Accessories,25,20,500.00
1,O1002,Ravi,South,Mouse,Accessories,20,20,400.00
2,O1019,Neha,East,Mouse,Accessories,18,20,360.00
3,O1003,Neha,West,Keyboard,Accessories,15,30,NaN
4,O1020,Amit,South,Keyboard,Accessories,14,30,420.00
5,O1013,Asha,North,Keyboard,Accessories,10,30,300.00
6,O1024,Amit,East,Monitor,Electronics,9,180,"1,620.00"
7,O1024,Amit,East,Monitor,Electronics,9,180,"1,620.00"
8,O1015,Neha,West,Chair,Furniture,8,90,720.00
9,O1004,Amit,East,Monitor,Electronics,7,180,"1,260.00"


## 17. Find highest and lowest Revenue

Use `max()` and `min()`; `idxmax()` and `idxmin()` identify the corresponding rows.

In [39]:
max_revenue = df['Revenue'].max()
max_revenue

3000.0

In [40]:
df['Revenue'].agg(['min', 'max'])

min     200.00
max   3,000.00
Name: Revenue, dtype: float64

In [41]:
# Which rows have max revenue values - 15th row
df.loc[df['Revenue'].idxmax()]

OrderID              O1016
Salesperson           Amit
Region               North
Product             Laptop
Category       Electronics
Units                    6
UnitPrice              500
Revenue           3,000.00
Name: 15, dtype: object

In [42]:
maxrev_region = df.loc[df['Revenue'].idxmax(), 'Region']
maxrev_region

'North'

In [43]:
df[df['Revenue'] == df['Revenue'].max()]

,OrderID,Salesperson,Region,Product,Category,Units,UnitPrice,Revenue
15,O1016,Amit,North,Laptop,Electronics,6,500,"3,000.00"


## 18. Calculate each row's percentage contribution to total Revenue

Divide each row's Revenue by total Revenue and multiply by 100.

In [44]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25 entries, 0 to 24
Data columns (total 8 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   OrderID      25 non-null     object 
 1   Salesperson  25 non-null     object 
 2   Region       25 non-null     object 
 3   Product      25 non-null     object 
 4   Category     25 non-null     object 
 5   Units        25 non-null     int64  
 6   UnitPrice    25 non-null     int64  
 7   Revenue      23 non-null     float64
dtypes: float64(1), int64(2), object(5)
memory usage: 1.7+ KB


In [45]:
df['Revenue_Perc'] = round(100 * df['Revenue'] / df['Revenue'].sum(), 2)
df.head()

,OrderID,Salesperson,Region,Product,Category,Units,UnitPrice,Revenue,Revenue_Perc
0,O1001,Asha,North,Laptop,Electronics,5,500,"2,500.00",11.47
1,O1002,Ravi,South,Mouse,Accessories,20,20,400.00,1.84
2,O1003,Neha,West,Keyboard,Accessories,15,30,NaN,NaN
3,O1004,Amit,East,Monitor,Electronics,7,180,"1,260.00",5.78
4,O1005,Asha,North,Printer,Office,3,220,660.00,3.03


In [46]:
## method 2: take sum of all row values in a column, divide each column value by Total Revenue
df['Total_Revenue'] = df['Revenue'].sum()
df['Revenue_Perc2'] = round(100 * df['Revenue'] / df['Total_Revenue'], 2)
df.head()

,OrderID,Salesperson,Region,Product,Category,Units,UnitPrice,Revenue,Revenue_Perc,Total_Revenue,Revenue_Perc2
0,O1001,Asha,North,Laptop,Electronics,5,500,"2,500.00",11.47,"21,790.00",11.47
1,O1002,Ravi,South,Mouse,Accessories,20,20,400.00,1.84,"21,790.00",1.84
2,O1003,Neha,West,Keyboard,Accessories,15,30,NaN,NaN,"21,790.00",NaN
3,O1004,Amit,East,Monitor,Electronics,7,180,"1,260.00",5.78,"21,790.00",5.78
4,O1005,Asha,North,Printer,Office,3,220,660.00,3.03,"21,790.00",3.03


## 19. Use IF logic to label Revenue as High or Low

Use `np.where()` for vectorised IF logic.

In [47]:
df['Revenue_Level'] = np.where(df['Revenue'] >= 1000, 'High', 'Low')

In [48]:
df[['OrderID', 'Revenue', 'Revenue_Level']].assign(compValue = 1000, Diff= lambda d:d['Revenue']- d['compValue']).head()

,OrderID,Revenue,Revenue_Level,compValue,Diff
0,O1001,"2,500.00",High,1000,"1,500.00"
1,O1002,400.00,Low,1000,-600.00
2,O1003,NaN,Low,1000,NaN
3,O1004,"1,260.00",High,1000,260.00
4,O1005,660.00,Low,1000,-340.00


In [49]:
df[['OrderID', 'Revenue', 'Revenue_Level']].assign(compValue = 1000, Diff= lambda d:d['Revenue']- d['compValue']).style.apply( lambda s: ['background-color: lightgreen' if v == 'High'
else 'background-color: salmon' if v == 'Low'   else '' for v in s],   subset=['Revenue_Level'])

,OrderID,Revenue,Revenue_Level,compValue,Diff
0,O1001,2500.000000,High,1000,1500.000000
1,O1002,400.000000,Low,1000,-600.000000
2,O1003,nan,Low,1000,nan
3,O1004,1260.000000,High,1000,260.000000
4,O1005,660.000000,Low,1000,-340.000000
5,O1006,nan,Low,1000,nan
6,O1007,540.000000,Low,1000,-460.000000
7,O1008,400.000000,Low,1000,-600.000000
8,O1009,500.000000,Low,1000,-500.000000
9,O1010,720.000000,Low,1000,-280.000000


## 20. Protect formula cells — Python equivalent

- A DataFrame has no editable formula cells. Protect the original data by keeping an untouched deep copy and creating derived columns in a separate result DataFrame.
- A deep copy of a DataFrame creates a completely independent copy of the data and index. Changes made to the copied DataFrame do not affect the original DataFrame.

In [50]:
df_original = df.copy(deep=True)
df_original.head()

,OrderID,Salesperson,Region,Product,Category,Units,UnitPrice,Revenue,Revenue_Perc,Total_Revenue,Revenue_Perc2,Revenue_Level
0,O1001,Asha,North,Laptop,Electronics,5,500,"2,500.00",11.47,"21,790.00",11.47,High
1,O1002,Ravi,South,Mouse,Accessories,20,20,400.00,1.84,"21,790.00",1.84,Low
2,O1003,Neha,West,Keyboard,Accessories,15,30,NaN,NaN,"21,790.00",NaN,Low
3,O1004,Amit,East,Monitor,Electronics,7,180,"1,260.00",5.78,"21,790.00",5.78,High
4,O1005,Asha,North,Printer,Office,3,220,660.00,3.03,"21,790.00",3.03,Low


## Additional useful analysis

In [51]:
# Active Variables
%whos

Variable                 Type             Data/Info
---------------------------------------------------
NamespaceMagics          MetaHasTraits    <class 'IPython.core.magi<...>mespace.NamespaceMagics'>
Path                     type             <class 'pathlib.Path'>
asha_average_units       float64          8.333333333333334
avg_revenue              float64          947.3913043478261
collections              module           <module 'collections' fro<...>collections/__init__.py'>
df                       DataFrame        Shape: (25, 12)
df_original              DataFrame        Shape: (25, 12)
df_revenue               DataFrame        Shape: (25, 8)
duplicate_orders         DataFrame        Shape: (2, 8)
electronics_revenue      float64          15120.0
get_ipython              function         <function get_ipython at 0x1060c0e00>
highlight_high_revenue   function         <function highlight_high_revenue at 0x14a2cc360>
islice                   type             <class 'itertools.islic

In [52]:
salesperson_summary = df.groupby('Salesperson', as_index=False).agg( Orders=('OrderID', 'count'),  Units=('Units', 'sum'),  \
    Revenue=('Revenue', 'sum'), Average_Revenue=('Revenue', 'mean')).sort_values('Revenue', ascending=False)
salesperson_summary

,Salesperson,Orders,Units,Revenue,Average_Revenue
0,Amit,7,51,"9,200.00","1,314.29"
1,Asha,6,50,"5,260.00",876.67
2,Neha,6,54,"5,120.00","1,024.00"
3,Ravi,6,34,"2,210.00",442.00


In [53]:
category_summary = df.groupby('Category', as_index=False).agg(  Orders=('OrderID', 'count'), Units=('Units', 'sum'), \
                         Revenue=('Revenue', 'sum')).sort_values('Revenue', ascending=False)
category_summary

,Category,Orders,Units,Revenue
1,Electronics,10,54,"15,120.00"
2,Furniture,6,24,"2,710.00"
0,Accessories,6,102,"1,980.00"
3,Office,3,9,"1,980.00"


## Consolidated answer summary

In [54]:
total_records = len(df)
total_revenue = df['Revenue'].sum()
avg_revenue = df['Revenue'].mean()
region_revenue = df.groupby('Region', as_index=False)['Revenue'].sum()
top_region = region_revenue.loc[region_revenue['Revenue'].idxmax()]
maxrev_region = df.groupby('Region', as_index=False)['Revenue'].sum()
category_revenue = df.groupby('Category', as_index=False)['Revenue'].sum()
top_category = category_revenue.loc[category_revenue['Revenue'].idxmax()]
max_revenue = df['Revenue'].max()
min_revenue = df['Revenue'].min()
dup_orders = df[df['OrderID'].duplicated(keep=False)]
dup_order = dup_orders['OrderID'].iloc[0] if not dup_orders.empty else None
dup_revenue = dup_orders['Revenue'].sum() if not dup_orders.empty else 0

In [55]:
measures =['Total Records','Total Revenue','Average Revenue','Region with Max Revenue']
answer = [total_records, total_revenue, avg_revenue, maxrev_region]
ansSummary = pd.DataFrame({"Measures":measures, 'Answer':answer})
ansSummary

,Measures,Answer
0,Total Records,25
1,Total Revenue,"21,790.00"
2,Average Revenue,947.39
3,Region with Max Revenue,"Region Revenue 0 East 6,960.00 1 North 7..."


In [56]:
print(f"- Total Revenue is **{total_revenue:,.0f}** and Average Revenue is **{avg_revenue:,.2f}**.")
print(f"- The {top_region['Region']} region has the highest regional Revenue at **{top_region['Revenue']:,.0f}**.")

- Total Revenue is **21,790** and Average Revenue is **947.39**.
- The North region has the highest regional Revenue at **7,580**.


## Key observations - with NA values
- Total Revenue is **21790** and Average Revenue is **947.30**.
- The North region has the highest regional Revenue at **7,580**.

## End of the exercise